cdap-docs/developers-manual/source/building-blocks/workflows.rst

```diff
@@ -86,12 +94,206 @@ implement the ``run()`` method::
+- alter the job configuration based on a key in the token; for example, set a different
+  mapper/reducer class or a different input/output dataset for a Spark or MapReduce program.
+  
+The API is intended to allow appropriate action to be taken in response to the token, including
+logging and modifying the conditional execution of the workflow based on the token.
+
+Once a run is completed, you can query the tokens from past workflow runs for analyses that
+determine which node was executed more frequently and when. You can retrieve the token values
+that were added by a specific node in the workflow to debug the flow of execution.
+
+Scope
+-----
+Two scopes |---| *System* and *User* |---| are provided for workflow keys. CDAP adds keys
+(such as MapReduce counters) under the *System* scope. Keys added by user programs are
+stored under the *User* scope.
+
+Putting and Getting Token Values
+--------------------------------
+When a value is put into a token, it is stored under a specific key. Both keys and their
+corresponding values must be non-null. The token stores additional information about the 
+context in which the key is being set, such as the unique name of the workflow node. 
+
+To put a value into a token, first obtain access to the token from the workflow context, and
+then set a value for a specific key, as shown in this example from a MapReduce mapper::
+
+  public static class MyVerifier extends Mapper<LongWritable, Text, Text, NullWritable> {
+    public void map(LongWritable key, Text value, Context context)
+      throws IOException, InterruptedException {
+      String profile;
+      WorkflowToken workflowToken = context.getWorkflowToken();
+      if (workflowToken != null) {
+        if (value != null and value.toString().equals("BuildProductProfile")) {
+          profile = "1";
+        } else {
+          profile = "0";
+        }
+        workflowToken.put("BuildProductProfile", profile);
+      }
+    }
+  }
+
+**Note:** The test of ``workflowToken != null`` is only required because this Mapper could
+be used outside of a workflow. When run from within a workflow, the token is guaranteed to
+be non-null.
+
+The `WorkflowToken Java API 
+<../../reference-manual/javadocs/co/cask/cdap/api/workflow/WorkflowToken.html)>`__
+includes methods for getting values for different keys, scopes, and nodes. The same
+key can be added to the workflow by different nodes, and there are methods to return a map of those
+key-value pairs. Convenience methods allow the putting and getting of non-string values
+through the use of the class Value.
+
+Spark Accumulators and Workflow Tokens
+--------------------------------------
+`Spark Accumulators <https://spark.apache.org/docs/latest/programming-guide.html#accumulators-a-nameaccumlinka>`__ 
+can be accessed through the SparkContext, and used with workflow tokens. This allows the 
+values in the accumulators to be accessed through workflow tokens and vice-versa.
```

Comment:
This is a little misleading. The vice-versa is not true. Also, for 3.1, users will have to manually add Accumulators to the workflow token. Example [here](https://github.com/caskdata/cdap/pull/3307/files#diff-aad4b0a2c7724edbf7d46f46590cd421R81)